components/web_discovery/browser/server_config_loader.cc

```diff
@@ -0,0 +1,446 @@
+      quorum_response_info->headers->response_code() != 200) {
+    VLOG(1) << "Failed to fetch server config";
+    return false;
+  }
+
+  auto collector_parsed_json = base::JSONReader::ReadAndReturnValueWithError(
+      *collector_response_body, base::JSON_PARSE_RFC);
+  auto quorum_parsed_json = base::JSONReader::ReadAndReturnValueWithError(
+      *quorum_response_body, base::JSON_PARSE_RFC);
+
+  if (!collector_parsed_json.has_value() || !quorum_parsed_json.has_value()) {
+    VLOG(1) << "Failed to parse server config json";
+    return false;
+  }
+
+  const auto* collector_root = collector_parsed_json.value().GetIfDict();
+  const auto* quorum_root = quorum_parsed_json.value().GetIfDict();
+  if (!collector_root || !quorum_root) {
+    VLOG(1) << "Failed to parse server config: not a dict";
+    return false;
+  }
+
+  const auto min_version = collector_root->FindInt(kMinVersionFieldName);
+  if (min_version && *min_version > kCurrentVersion) {
+    VLOG(1) << "Server minimum version is higher than current version, failing";
+    return false;
+  }
+
+  auto config = std::make_unique<ServerConfig>();
+
+  const auto* group_pub_keys = collector_root->FindDict(kGroupPubKeysFieldName);
+  if (!group_pub_keys) {
+    VLOG(1) << "Failed to retrieve groupPubKeys from server config";
+    return false;
+  }
+  const auto* pub_keys = collector_root->FindDict(kPubKeysFieldName);
+  if (!pub_keys) {
+    VLOG(1) << "Failed to retrieve pubKeys from server config";
+    return false;
+  }
+  const auto* source_map = collector_root->FindDict(kSourceMapFieldName);
+  const auto* source_map_actions =
+      source_map ? source_map->FindDict(kSourceMapActionsFieldName) : nullptr;
+  if (!source_map_actions) {
+    VLOG(1) << "Failed to retrieve sourceMap from server config";
+    return false;
+  }
+
+  const auto* location = quorum_root->FindString(kLocationFieldName);
+  if (location && kAllowedReportLocations.contains(*location)) {
+    config->location = *location;
+  } else {
+    config->location = kOmittedLocationValue;
+  }
+
+  config->group_pub_keys = ParseKeys(*group_pub_keys);
+  config->pub_keys = ParseKeys(*pub_keys);
+  config->source_map_actions = ParseSourceMapActionConfigs(*source_map_actions);
+
+  last_loaded_server_config_ = std::move(config);
+  config_callback_.Run();
+  return true;
+}
+
+void ServerConfigLoader::LoadStoredPatterns() {
+  sequenced_task_runner_->PostTaskAndReplyWithResult(
+      FROM_HERE, base::BindOnce(&ReadPatternsFile, patterns_path_),
+      base::BindOnce(&ServerConfigLoader::OnPatternsFileLoaded,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ServerConfigLoader::OnPatternsFileLoaded(
+    std::optional<std::string> patterns_json) {
+  if (!patterns_json) {
+    VLOG(1) << "Failed to load local patterns file";
+    local_state_->ClearPref(kPatternsRetrievalTime);
+    SchedulePatternsRequest();
+    return;
+  }
+  auto parsed_patterns = ParsePatterns(*patterns_json);
```

Comment:
`Parse` should happen in background. Please consider always use background threads to parse JSONs.